File: 160.intersection-of-two-linked-lists.java
/*
 * @lc app=leetcode id=160 lang=java
 *
 * [160] Intersection of Two Linked Lists
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 * int val;
 * ListNode next;
 * ListNode(int x) {
 * val = x;
 * next = null;
 * }
 * }
 */
public class Solution {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        ListNode iterA = headA, iterB = headB;
        int sizeA = 0, sizeB = 0;
        while (iterA != null) {
            sizeA += 1;
            iterA = iterA.next;
        }
        while (iterB != null) {
            sizeB += 1;
            iterB = iterB.next;
        }
        iterA = headA;
        iterB = headB;
        if (sizeA > sizeB) {
            int n = sizeA - sizeB;
            while (n != 0) {
                iterA = iterA.next;
                n -= 1;
            }
        } else {
            int n = sizeB - sizeA;
            while (n != 0) {
                iterB = iterB.next;
                n -= 1;
            }
        }
        while (iterA != null) {
            if (iterA == iterB) {
                return iterA;
            }
            iterA = iterA.next;
            iterB = iterB.next;
        }
        return null;
    }
}
// @lc code=end

File: 160.intersection-of-two-linked-lists.cpp
/*
 * @lc app=leetcode id=160 lang=cpp
 *
 * [160] Intersection of Two Linked Lists
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        int sizeA = size(headA), sizeB = size(headB);
        ListNode *longer = (sizeA > sizeB) ? headA : headB;
        ListNode *shorter = (longer == headA) ? headB : headA;
        int diff = (sizeA > sizeB) ? sizeA - sizeB : sizeB - sizeA;
        while (diff != 0) {
            longer = longer->next;
            diff -= 1;
        }
        while (longer != nullptr && longer != shorter) {
            longer = longer->next;
            shorter = shorter->next;
        }
        return longer;
    }
    int size(ListNode *head) {
        int sz = 0;
        while (head != nullptr) {
            sz += 1;
            head = head->next;
        }
        return sz;
    }
};
// @lc code=end
